### PR TITLE
mon: Improvements to slow heartbeat health messages

### PR DESCRIPTION
**Pull request by dzafman**
Include crush parentage for each osd

## Checklist
- [x] References tracker ticket
- [x] Updates documentation if necessary
- [x] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by dzafman** [comment]
@liewegas Changes per your suggestions in December 4 e-mail, then documented in https://tracker.ceph.com/issues/43126

File: src/mon/PGMap.cc [hunk 1]
```diff
@@ -2363,6 +2363,26 @@ void PGMap::dump_pool_stats_and_io_rate(int64_t poolid, const OSDMap &osd_map,
...
 
+// Get crush parentage for an osd (skip root)
+set<std::string> PGMap::osd_parentage(const OSDMap& osdmap, int id) const
+{
+  set<std::string> reporters_by_subtree;
+  // XXX: We really want whatever has OSDs
+  auto reporter_subtree_level = g_conf().get_val<string>("mon_osd_reporter_subtree_level");
+
+  auto loc = osdmap.crush->get_full_location(id);
+  for (map<string,string>::iterator p = loc.begin(); p != loc.end(); ++p) {
```
Comment (tchaikov) [inline src/mon/PGMap.cc]: could put
```c++
for (auto& [parent_bucket_type, parent_id] : loc) {
  if (parent_bucket_type == "root") {
    continue;
  } else if (parent_bucket_type != reporter_subtree_level) {
    reporters_by_subtree.insert(parent_id);
  }
}
```

File: src/mon/PGMap.cc [hunk 2]
```diff
@@ -2363,6 +2363,26 @@ void PGMap::dump_pool_stats_and_io_rate(int64_t poolid, const OSDMap &osd_map,
   }
 }
 
+// Get crush parentage for an osd (skip root)
+set<std::string> PGMap::osd_parentage(const OSDMap& osdmap, int id) const
+{
+  set<std::string> reporters_by_subtree;
```
Comment (tchaikov) [inline src/mon/PGMap.cc]: is it possible that an OSD have the same parent for multiple times?

File: src/mon/PGMap.cc [hunk 3]
```diff
@@ -2363,6 +2363,26 @@ void PGMap::dump_pool_stats_and_io_rate(int64_t poolid, const OSDMap &osd_map,
...
+
+  auto loc = osdmap.crush->get_full_location(id);
+  for (map<string,string>::iterator p = loc.begin(); p != loc.end(); ++p) {
+    if (p->first == "root") continue;
+    if (p->first != reporter_subtree_level) {
+      reporters_by_subtree.insert(p->second);
+    }
+  }
+  for (auto s: reporters_by_subtree)
+	  dout(0) << __func__ << " " << s << dendl;
```
Comment (tchaikov) [inline src/mon/PGMap.cc]: presumably, this is not need since it's also printed by the next line.

File: src/mon/PGMap.cc [hunk 4]
```diff
@@ -2363,6 +2363,26 @@ void PGMap::dump_pool_stats_and_io_rate(int64_t poolid, const OSDMap &osd_map,
...
+  dout(0) << __func__ << " set " << reporters_by_subtree << dendl;
```
Comment (tchaikov) [inline src/mon/PGMap.cc]: can we use a higher debug level for this message? as to my understanding, this is not strictly an error.

Comment (dzafman) [reply src/mon/PGMap.cc]: mistake

File: src/mon/PGMap.cc (hunk 2 above)
Comment (dzafman) [reply src/mon/PGMap.cc]: Yes, but in real life these errors should be rare and I don't think it is worth caching the results for efficiency.

File: src/mon/PGMap.cc (hunk 4 above)
Comment (dzafman) [reply src/mon/PGMap.cc]: Code added for development testing, removed.

**Comment by dzafman** [comment]
Testing passed

**Pull request merged**